src/services/cleeng.auth.service.ts

```diff
@@ -0,0 +1,318 @@
+
+const AUTH_PERSIST_KEY = 'auth';
+
+type Tokens = {
+  accessToken: string;
+  refreshToken: string;
+};
+
+type MessageAction = 'refreshing' | 'resolved' | 'rejected';
+
+type MessageData = {
+  action: MessageAction;
+  tokens?: Tokens;
+};
+
+type JWTPayload = {
+  exp?: number;
+};
+
+const tokensSchema = object().shape({
+  accessToken: string().required(),
+  refreshToken: string().required(),
+});
+
+/**
+ * Validate the given input if it confirms to the tokenSchema
+ */
+function isValidTokens(candidate: unknown): candidate is Tokens {
+  return tokensSchema.isValidSync(candidate);
+}
+
+/**
+ * Given an JWT, return token expiration in milliseconds. Returns -1 when the token is invalid.
+ */
+function getTokenExpiration(token: string) {
+  try {
+    const decodedToken: JWTPayload = jwtDecode(token);
+
+    if (typeof decodedToken.exp === 'number') {
+      return decodedToken.exp * 1000;
+    }
+  } catch (error: unknown) {
+    // failed to parse the JWT string
+  }
+
+  return -1;
+}
+
+/**
+ * Persist the given token in the storage. Removes the token when the given token is `null`.
+ */
+async function persistInStorage(tokens: Tokens | null) {
+  if (tokens) {
+    await setItem(AUTH_PERSIST_KEY, JSON.stringify(tokens));
+  } else {
+    await removeItem(AUTH_PERSIST_KEY);
+  }
+}
+
+/**
+ * The AuthService is responsible for managing JWT access tokens and refresh tokens.
+ *
+ * Once an access token and refresh token is set, it will automatically refresh the access token when it is about to
+ * expire.
+ *
+ * It uses a PromiseQueue to prevent multiple instances refreshing the same token which fails.
+ *
+ * The Broadcaster ensures that all potential browser tabs are notified about the updated token. This prevents
+ * race-conditions when the user has multiple tabs open (but also very helpful while developing).
+ */
+class AuthService {
+  private readonly channel: Broadcaster<MessageData>;
+  private readonly queue = new PromiseQueue();
+  private isRefreshing = false;
+  private expiration = -1;
+  private sandbox = false;
+  tokens: Tokens | null = null;
+
+  /**
+   * The `logoutCallback` is a delegate that can be set to handle what should happen when the token is expired or not
+   * valid anymore. Since this service shouldn't care about the UI or Zustand state, a controller can use this delegate
+   * to update the UI and clear the logged in state. Alternatively, this could dispatch an event, but a delegate is
+   * easier to align with the InPlayer integration.
+   */
+  private logoutCallback: () => Promise<void> = () => Promise.resolve();
+
+  /**
+   * This function does the actual refresh of the access token by calling the `refreshToken` endpoint in the Cleeng
+   * service.
+   */
+  private getNewTokens: (tokens: Tokens) => Promise<Tokens | null> = async ({ refreshToken }) => {
+    try {
+      const { responseData: newTokens } = await accountService.refreshToken({ refreshToken }, this.sandbox);
```

Comment:
Is `refreshToken` used anywhere else? If not, can it be moved into auth service? It would be nice not to have circular dependency between auth service and account service.

Reply:
Nope, and you're entirely right. The auth service was initially setup as a generic service (that doesn't know about Cleeng), but now that it is coupled I moved the `refreshToken` function the cleeng.auth.service